Platform/OpenNtfsDxe/Index.c

```diff
@@ -0,0 +1,918 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+extern UINTN  mFileRecordSize;
+extern UINTN  mIndexRecordSize;
+STATIC UINT64 mBufferSize;
+
+STATIC
+VOID
+FreeNode (
+  IN NTFS_FILE  *Node    OPTIONAL,
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  if (Node != Context->RootNode) {
+    if (Node != NULL) {
+      FreePool (Node);
+    }
+  }
+}
+
+STATIC
+VOID
+PopElement (
+  IN FSHELP_CTX *Context
+  )
+{
+  STACK_ELEMENT *Element;
+
+  ASSERT (Context != NULL);
+
+  Element = Context->CurrentNode;
+  Context->CurrentNode = Element->Parent;
+  FreeNode (Element->Node, Context);
+  FreePool (Element);
+}
+
+STATIC
+VOID
+FreeStack (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  while (Context->CurrentNode != NULL) {
+    PopElement (Context);
+  }
+}
+
+STATIC
+VOID
+GoUpALevel (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  if (Context->CurrentNode->Parent == NULL) {
+    return;
+  }
+  PopElement (Context);
+}
+
+STATIC
+EFI_STATUS
+PushNode (
+  IN FSHELP_CTX       *Context,
+  IN NTFS_FILE        *Node,
+  IN FSHELP_FILETYPE  FileType
+  )
+{
+  STACK_ELEMENT *Next;
+
+  ASSERT (Context != NULL);
+  ASSERT (Node != NULL);
+
+  Next = AllocateZeroPool (sizeof (STACK_ELEMENT));
+  if (Next == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Next->Node = Node;
+  Next->Type = FileType & ~FSHELP_CASE_INSENSITIVE;
+  Next->Parent = Context->CurrentNode;
+  Context->CurrentNode = Next;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+GoToRoot (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  FreeStack (Context);
+  return PushNode (Context, Context->RootNode, FSHELP_DIR);
+}
+
+STATIC
+EFI_STATUS
+FindFileIter (
+  IN CHAR16           *Name,
+  IN FSHELP_FILETYPE  FileType,
+  IN NTFS_FILE        *Node,
+  IN FSHELP_ITER_CTX  *Context
+  )
+{
+  INTN Result;
+
+  ASSERT (Name != NULL);
+  ASSERT (Node != NULL);
+  ASSERT (Context != NULL);
+
+  if (FileType == FSHELP_UNKNOWN) {
+    return EFI_NOT_FOUND;
+  }
+
+  if ((FileType & FSHELP_CASE_INSENSITIVE) != 0) {
+    Result = OcStriCmp (Context->Name, Name);
+  } else {
+    Result = StrCmp (Context->Name, Name);
+  }
+
+  if (Result != 0) {
+    return EFI_NOT_FOUND;
+  }
+
+  *Context->FoundNode = Node;
+  *Context->FoundType = FileType;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+NtfsDirHook (
+  IN CHAR16           *Name,
+  IN FSHELP_FILETYPE  FileType,
+  IN NTFS_FILE        *Node,
+  IN EFI_FILE_INFO    *Info
+  )
+{
+  EFI_STATUS  Status;
+  INT64       *Index = (INT64 *) &Info->FileSize;
```

Comment:
Initialize variables later.

Reply:
Why is this cast? Also, dereference is before ASSERT